File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/messaging/MailService.java
package top.likoslupus.cellulosesz.api.messaging;

import java.util.Collection;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public interface MailService {

    CompletableFuture<Void> send(MailMessage message);

    CompletableFuture<Integer> sendAll(Collection<UUID> recipients, MailMessageFactory factory);

    CompletableFuture<List<MailMessage>> inbox(UUID recipient);

    CompletableFuture<Integer> unreadCount(UUID recipient);

    CompletableFuture<Void> markRead(UUID recipient, Collection<UUID> messageIds);

    CompletableFuture<Boolean> delete(UUID recipient, UUID messageId);

    CompletableFuture<Integer> clear(UUID recipient);

    CompletableFuture<Integer> purgeExpired(long now);

    @FunctionalInterface
    interface MailMessageFactory {

        MailMessage create(UUID recipient);

    }

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/messaging/MailService.kt
package top.likoslupus.cellulosesz.api.messaging

import java.util.*
import java.util.concurrent.CompletableFuture

public interface MailService {

    public fun send(message: MailMessage): CompletableFuture<Void>

    public fun sendAll(
        recipients: Collection<UUID>,
        factory: MailMessageFactory
    ): CompletableFuture<Int>

    public fun inbox(recipient: UUID): CompletableFuture<List<MailMessage>>

    public fun unreadCount(recipient: UUID): CompletableFuture<Int>

    public fun markRead(
        recipient: UUID,
        messageIds: Collection<UUID>
    ): CompletableFuture<Void>

    public fun delete(
        recipient: UUID,
        messageId: UUID
    ): CompletableFuture<Boolean>

    public fun clear(recipient: UUID): CompletableFuture<Int>

    public fun purgeExpired(now: Long): CompletableFuture<Int>

    public fun interface MailMessageFactory {

        public fun create(recipient: UUID): MailMessage

    }

}
